### PR TITLE
Custom Edit Context Views

### PR DESCRIPTION
**Pull request by natis1**
Adds custom edit context views and closes #7

The views added are "Armature Mode", "Skeleton Mode", "Muscle Mode", and "Skin Mode"

Additionally, there's a "Smart Object Mode" which functions as "Bone Edit Context" would but can be run on any object, not just bones.

Smart Object Mode will make visible any selected objects, plus any objects in any geometry nodes present on these selected objects.

### CODE REVIEW

File: scripts/ui/mode_selection.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+import bpy
+
```
Comment (Lathreas) [inline scripts/ui/mode_selection.py:8]: Please add a module docstring.

File: scripts/ui/mode_selection.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                obj.hide_set(True)
+        return {'FINISHED'}
+
+
+class ARF_OT_Skeleton_Mode(bpy.types.Operator):
+    bl_label = "Skeleton Mode"
+    bl_idname = "arf.skeleton_mode"
+
+    def execute(self, context):
+        for obj in context.scene.objects:
```
Comment (Lathreas) [inline scripts/ui/mode_selection.py:28]: Please make sure to check if `obj.ARF` exists before using it, with an assertion. `obj.ARF` should in principle always be present, hence asserting, but this way `mypy` won't complain and it allows us to verify that the code works as intended.
```suggestion
        for obj in context.scene.objects:
            assert hasattr(obj, 'ARF')
```

File: scripts/ui/__init__.py [hunk 3]
```diff
@@ -3,13 +3,17 @@
 """Contains all user interface elements for the Blender plugin."""
 
 from . import bone_properties
-
+from . import mode_selection
 
 def register_all():
     bone_properties.register()
+    mode_selection.register()
+    #object_type_selection.register()
```
Comment (Lathreas) [inline scripts/ui/__init__.py]: Please remove this commented code before merging (registering `object_type_selection` should go in a different PR).

File: scripts/ui/__init__.py [hunk 4]
```diff
@@ -3,13 +3,17 @@
...
     print("Registered scripts.ui")
 
 
 def unregister_all():
     bone_properties.unregister()
+    mode_selection.unregister()
+    #object_type_selection.unregister()
```
Comment (Lathreas) [inline scripts/ui/__init__.py]: Please remove this commented code before merging (registering `object_type_selection` should go in a different PR).

File: scripts/ui/mode_selection.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                    # For some reason input.default_value is a copy of the object and not the actual
+                    # object in the scene, so we need to get the actual object in the scene.
+                    inputobj = bpy.context.scene.objects.get(input.default_value.name)
+                    inputobj.hide_set(False)
+                    # We also need to set every parent to also be visible
+                    parent = inputobj.parent
+                    while (parent != None):
+                        parent.hide_set(False)
+                        parent = parent.parent
+
```
Comment (Lathreas) [inline scripts/ui/mode_selection.py]: Ah, I see you are looping over the modifier list. Sadly that has the effect of displaying the objects that directly modify the bones, which may not be the objects that the user can edit (instead being intermediary objects). These objects can be rather large and clutter the viewport, due to their mathematical requirement of being large, so it's best if they stay hidden. The preferred way to display the correct objects is to loop over `selection.ARF.bone_properties.geometry_constraints` and then getting `constraint_object` from that. e.g.:

```python
for selection in selected:
    for geometry_constraint in selection.ARF.bone_properties.geometry_constraints:
        geometry_constraint.constraint_object.hide_set(False)
```

Let me send you a .blend file where an object has those geometry_constraints properly filled in so that you can test the code!

File: scripts/ui/mode_selection.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
+import bpy
```
Comment (Lathreas) [inline scripts/ui/mode_selection.py:7]: Please add `#!/usr/bin/python3` at the start of the file, that helps certain operating systems recognize it as a Python3 file.

**Review by Lathreas — CHANGES_REQUESTED** [review]
Thank you for the Pull Request! ^.=.^ I've done a review. The code looks good, and the design is sound. The only code logic change I propose is in ARF_OT_Smart_Object_Mode.

There are a few style issues that should be addressed too, and then it can be merged. A bit nitpicky, but that keeps the code clean of course. I realize that the Lathreas/user-interface also doesn't yet pass the style review, so consider this style guide to be necessary for when it all gets merged into main. We might as well get that done now, so here it goes!

 - Make sure your comments don't exceed the 88 character line limit. Feel free to split them into multiple lines if necessary.
 - Make sure to end comments with a period (.) or colon (:) .
 - Please add (simple) docstrings to the classes. No need to be fancy, for example adding something like `"""Sets the ARF mode to Skeleton Mode."""` (for the skeleton mode operator class) is enough for most of the classes here.
 - Please add a module docstring to explain what the file is for.
 
 Thanks again for your work!

**Comment by natis1** [comment]
I believe the newest commit should resolve all of the issues that you brought up. I tested with the startup.blend you sent me and it did seem to work as expected.

**Review by Lathreas — APPROVED** [review]
Awesome!

LGTM.
